### PR TITLE
Support HTTP POST + gzipped streams for Wavefront/Graphite/OpenTSDB 

### PR DESCRIPTION
**Pull request by basilisk487**

### CODE REVIEW

File: proxy/src/main/java/com/wavefront/agent/PlainTextOrHttpFrameDecoder.java [hunk 1]
```diff
@@ -66,13 +74,23 @@ protected void decode(final ChannelHandlerContext ctx,
...
-      pipeline.addLast("decoder", new HttpRequestDecoder());
-      pipeline.addLast("encoder", new HttpResponseEncoder());
-      pipeline.addLast("aggregator", new HttpObjectAggregator(16 * 1024 * 1024));
-      pipeline.addLast("deflate", new HttpContentCompressor());
-      pipeline.addLast("handler", this.handler);
+
+    if (detectGzip && isGzip(firstByte, secondByte)) {
+      logger.fine("Inbound gzip stream detected");
+      pipeline
+          .addLast("gzipdeflater", ZlibCodecFactory.newZlibEncoder(ZlibWrapper.GZIP))
```
Comment (panghy) [inline proxy/src/main/java/com/wavefront/agent/PlainTextOrHttpFrameDecoder.java]: should any of the encoders gzip and http be cached?

File: proxy/src/main/java/com/wavefront/agent/PlainTextOrHttpFrameDecoder.java [hunk 2]
```diff
@@ -66,13 +74,23 @@ protected void decode(final ChannelHandlerContext ctx,
...
+          .addLast("gzipinflater", ZlibCodecFactory.newZlibDecoder(ZlibWrapper.GZIP))
+          .addLast("unificationB", new PlainTextOrHttpFrameDecoder(handler, false));
+    } else if (isHttp(firstByte, secondByte)) {
+      logger.fine("Switching to HTTP protocol");
+      pipeline
+          .addLast("decoder", new HttpRequestDecoder())
```
Comment (panghy) [inline proxy/src/main/java/com/wavefront/agent/PlainTextOrHttpFrameDecoder.java]: ditto

File: proxy/src/main/java/com/wavefront/agent/PortUnificationHandler.java [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+
+import static io.netty.handler.codec.http.HttpHeaders.Names.CONNECTION;
+import static io.netty.handler.codec.http.HttpHeaders.Names.CONTENT_LENGTH;
+import static io.netty.handler.codec.http.HttpHeaders.Names.CONTENT_TYPE;
+import static io.netty.handler.codec.http.HttpVersion.HTTP_1_1;
+
+/**
+ * This class handles an incoming message of either String or FullHttpRequest type.  All other types are ignored. This
+ * will likely be passed to the PlainTextOrHttpFrameDecoder as the handler for messages.
+ *
```
Comment (panghy) [inline proxy/src/main/java/com/wavefront/agent/PortUnificationHandler.java]: add author

File: proxy/src/main/java/com/wavefront/agent/PortUnificationHandler.java [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+  }
+
+  @Override
+  public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
+    logWarning("Handler failed", cause, ctx);
+  }
+
+  @Override
+  protected void channelRead0(final ChannelHandlerContext ctx,
+                              final Object message) {
```
Comment (panghy) [inline proxy/src/main/java/com/wavefront/agent/PortUnificationHandler.java]: should indent only

File: proxy/src/main/java/com/wavefront/agent/PortUnificationHandler.java [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+        ctx.channel().close();
+      }
+    }
+  }
+
+  /**
+   * Writes an HTTP response.
+   */
+  protected void writeHttpResponse(ChannelHandlerContext ctx,
+                                   HttpResponseStatus status,
```
Comment (panghy) [inline proxy/src/main/java/com/wavefront/agent/PortUnificationHandler.java]: ditto, check code style

File: proxy/src/main/java/com/wavefront/agent/PortUnificationHandler.java [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+                                   Object contents,
+                                   boolean keepAlive) {
+    FullHttpResponse response;
```
Comment (panghy) [inline proxy/src/main/java/com/wavefront/agent/PortUnificationHandler.java]: final?

File: proxy/src/main/java/com/wavefront/agent/PlainTextOrHttpFrameDecoder.java (hunk 1 above)
Comment (basilisk487) [reply proxy/src/main/java/com/wavefront/agent/PlainTextOrHttpFrameDecoder.java]: In Netty everything http and gzip-related is not stateless, unfortunately, so these objects can't be shared.

**Comment by panghy** [comment]
@basilisk487 what's the status on this PR?

**Comment by basilisk487** [comment]
@panghy Added unit tests - these should be a good start

**Comment by panghy** [comment]
Thanks

**Pull request merged**